### PR TITLE
Lupus import

### PR DESCRIPTION
**Pull request by andrewram4287**

### CODE REVIEW

**Comment by pascoesean** [comment]
@aarmey I think this pull request should work to merge what I have been working on (I ran into a few git issues [gitssues?] earlier in the week and had to copy/delete previous files from the confusingly named `LupusImport` branch. 

Here's a quick overview of the 3 files that I'm including:

 **_figureS1_**: initial visualization of expression value distributions used to understand normalization
 **_figureS2_**: initial run of pf2, with factor plots [if you want to try running pf2 on the whole dataset this would probably be a good place to do it, I can leave a comment there too]
 **_figureS3_**: investigation of cell type contribution to cell state by UMAP and with violin plots derived from projections matrix

File: sccp/figures/figureS2.py [hunk 1]
```diff
@@ -1,133 +1,150 @@
...
+    # REMOVE THIS SECTION TO HAVE ALL INDIVIDUALS INCLUDED (and n_rand arg in function def)
+    #@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+    np.random.seed(42)
+    SLE_patients = X.obs["ind_cov"][X.obs["SLE_status"] == "SLE"].to_numpy()
+    healthy_patients = X.obs["ind_cov"][X.obs["SLE_status"] == "Healthy"].to_numpy()
+    random_patients = np.random.choice(SLE_patients, int(n_rand/2), replace=False).tolist() + np.random.choice(healthy_patients, int(n_rand/2), replace=False).tolist()
+
+    X = X[X.obs['ind_cov'].isin(random_patients), :]
+    #@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+    #@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
```
Comment (pascoesean) [inline sccp/figures/figureS2.py:55]: @aarmey this can be removed alongside the `n_rand` keyword argument to run the full dataset

File: sccp/figures/figureS1.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+"""
```
Comment (aarmey) [inline sccp/figures/figureS1.py:1]: Remove this from the PR. You can keep it as its useful code, but this file does not create a figure. Alternatively, this could be output within a file that creates one of the other figures, if you find it would be very useful to keep around.

File: sccp/figures/figureS2.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+import matplotlib.pyplot as plt
+import scipy.cluster.hierarchy as sch
+import tensorly as tl
+
+
+####################################################################################
+## LOAD IN DATA -- CAN BE MOVED TO imports/scRNA 
+####################################################################################
+
+def load_lupus_data(third_axis= "ind_cov", n_rand = 100):
```
Comment (aarmey) [inline sccp/figures/figureS2.py:31]: This function can go into `imports/scRNA.py` so that it is defined once for all figures.

File: sccp/figures/figureS2.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    obsV = X.obs_vector('ind_cov')
+    sgUnique, sgIndex = np.unique(obsV, return_inverse=True)
+
+    ann_data_objects = [X[sgIndex == sgi, :] for sgi in range(len(sgUnique))]
+
+    X = anndata.concat(ann_data_objects, axis = 0)
+
+    # select n_rand random patients; evenly split between lupus and SLE
+    #@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+    # REMOVE THIS SECTION TO HAVE ALL INDIVIDUALS INCLUDED (and n_rand arg in function def)
```
Comment (aarmey) [inline sccp/figures/figureS2.py:46]: Doesn't it make more sense to keep all patients, but downsample the cells? I thought I had that in the code I passed along. That way, any downstream analysis isn't affected.

I suppose maybe you did this because of that one patient with fewer cells. Another way to downsample the dataset would be to remove cells only from the patients with the most. e.g., set an upper bound that patients with more than N cells have their data cut to N cells.

File: sccp/figures/figureS2.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    lut = {'SLE': 'c', 'Healthy': 'm'}
+    row_colors = status['SLE_status'].map(lut)
+
+    assert np.all(np.isfinite(X.X.data)) # this should be true
+
+    return tensorFy(X, third_axis), cell_types, row_colors
+
+####################################################################################
+
+def plotLupusFactors(factors, data: Pf2X, row_colors, axs, reorder=tuple(), trim=tuple()):
```
Comment (aarmey) [inline sccp/figures/figureS2.py:73]: How is this function different from the existing plotFactors function? Ideally we would reuse that code.

File: sccp/figures/figureS3.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+from sklearn.decomposition import PCA
+from matplotlib import pyplot
+
+
+
+####################################################################################
+## LOAD IN DATA -- CAN BE MOVED TO imports/scRNA 
+####################################################################################
+
+def load_lupus_data(third_axis= "ind_cov", n_rand = 100):
```
Comment (aarmey) [inline sccp/figures/figureS3.py:32]: See comment above about this function.

**Review by aarmey — CHANGES_REQUESTED** [review]
@pascoesean Looks good! I've added some structural suggestions here.

**Comment by aarmey** [comment]
@pascoesean it looks like it will require some more complicated shuffling between the CPU and GPU to scale to much larger datasets. I plan to implement this, but it will likely be a week or two, so I would proceed with my suggestion above about using all patients and downsampling those with a large number of cells. I'll keep you posted on the Pf2 algorithm changes.

**Comment by aarmey** [comment]
@andrewram4287 it looks like the build is still failing due to an error in Fig. 9. Can you take a look?

File: sccp/figures/figureS2.py (hunk 4 above)
Comment (pascoesean) [reply sccp/figures/figureS2.py:46]: Yeah downsampling cells ran into the issue where we were using a number of cells that wasn't much bigger than the number of cell states. If we can do every 4th cell (instead of every 10th) this would probably be fine, I can also definitely try downsampling from those with the most cells. I think our biggest concern when selecting cells, rather than patients, was that cell types that are less common but present in all patients might not be included or might be included with an unfair weight? But with this many observations that might not make a difference -- @andrewram4287  let me know if you have other thoughts on that too.

Here's a histogram of counts per patient if that helps to get an idea of what's going on:



image

File: sccp/figures/figureS2.py (hunk 5 above)
Comment (pascoesean) [reply sccp/figures/figureS2.py:73]: `PlotLupusFactors()` has a `row_colors` argument that sets the key to make a bar on the left hand side of the A matrix plot to visualize which patients were SLE/healthy. I can definitely add this as an optional argument to `PlotFactors()` that defaults to None (since each of the drugs are different in the Thomson lab dataset, we wouldn't necessarily need it for that); mostly I was just hesitant to change other files

Comment (aarmey) [reply sccp/figures/figureS2.py:73]: Feel free to add it to the master function!

File: sccp/figures/figureS2.py (hunk 4 above)
Comment (aarmey) [reply sccp/figures/figureS2.py:46]: Why don't you do every fourth cell for now, since that should be simple to implement, and then we can remove that when I address the scaling? Even every third cell might fit into memory just fine.

You're right to be concerned about downsampling in any form. Indeed—you could lose rare populations. However, it's probably worse to subsample the patients, because you only have a couple hundred of those. Just like more cells is an opportunity to see rare populations, each patient is an opportunity for the model to build (or refute) its evidence for a pattern being associated with disease.

**Comment by pascoesean** [comment]
@andrewram4287 this is my best attempt at a review request

**Review by aarmey — APPROVED** [review]
This all looks good to me.
